### ISSUE TITLE
Transaction fee isn't collected by the slot leader

### ISSUE THREAD

**Issue by mvines** [solana-labs/solana#2623]
The fee for a Transaction is currently debited
https://github.com/solana-labs/solana/blob/1c0758e3bdf4fb897d063294aec9793e947db1da/src/accounts.rs#L152
but it is not credited anywhere, effectively burning the fee.

Instead the Transaction fee should be credited to the slot leader's staking account.

cc: #1527

**Comment by garious** [comment]
@pgarg66, how exactly is this blocked?

**Comment by pgarg66** [comment]
This will touch bank, and accounts. I think it'll cause issues with bank forking stuff if done in parallel.

**Comment by pgarg66** [comment]
Should the fee be credited to voting account keypair or leader's keypair created in fullnode new?

**Comment by pgarg66** [comment]
Also, my current approach is that each node finds the current leader and credits the fee to leader's account. But can this lead to different balance for the leader's account on different nodes?

Also, if the fork does not get merged to the root, the leader will still collect the transaction fee. Is that the right behavior?

**Comment by garious** [comment]
@pgarg66 The leader's keypair.   The transaction fee should be credited after `commit_transactions()` in the bank. That should cause it to be the same from both BankingStage and ReplayStage.

**Comment by pgarg66** [comment]
What can we do for this use case?

At boot up, a node runs ```process_ledger()```, which processes all entries in the ledger thru bank. Can the leader schedule be out of bound for a really old entry? If so, the node wouldn't know which leader account should be credited for the transaction.

Or, is there any other code flow through which the node accounts are getting updated/synchronized?

**Comment by garious** [comment]
The leader scheduler is now explicitly passed to the `process_ledger()` functions. It's height is updated after every `register_tick()`. https://github.com/solana-labs/solana/blob/master/src/blocktree_processor.rs#L122

Before that change, the leader scheduler was updated at the same time, but it was done implicitly, from a LeaderScheduler reference that was passed into the Bank's constructor.

**Comment by garious** [comment]
You should write some times to ensure your assumptions are valid (and stay valid). I'd love to rip the leader scheduler out of there and instead pass a leader *schedule*, that'd be valid for one epoch at a time.  You'd run `process_block(&entries, &mut leader_schedule)` and it'd pop a slot off the schedule after the block boundary.

**Comment by pgarg66** [comment]
Capturing the discussion from discord on this issue.

```
pgarg: Do we need to collect the transaction fee in TPU (banking_stage) as well,
or just in blocktree_processor (in replay stage)?
Maybe better question is, is a leader's replay stage processing it's own transactions as well?

greg: @pgarg, the ReplayStage should not process its own transactions.
greg: Hence, the easiest place to collect fees would be the same place ReplayStage
would vote on a block, which is supposed to be the same  on  both leader and validator.

pgarg: So, leader's banking_stage needs to tracks it's fee.
And apply the credit to leader's account at the tick height at which validators will vote.
In a validator, the blocktree_processor will track the fee,
and apply on tick height when there's a vote. Does this look correct?

greg: Shouldn't need to do anything in banking_stage.
If you do, it's because TPU isn't getting it's fork to the TVU to vote on.

pgarg: I am a bit confused. The ReplayStage in TVU processes the entries and votes.
If the ReplayStage is not processing it's own transactions (when it's a leader node),
how will it vote on it's own fork?
Does blocktree_processor::process_entries() skip over it's own entries?

greg: @pgarg, yes, ReplayStage code is terrible.
It needs to be cleaned up so that you can do what you described.
greg: Or if you're not comfortable with that,  you could hack it  up further
and filter out entries within its own leader slot.
greg: that'd make a terrible thing worse though...

sakridge: it already has that logic I believe
sakridge: actually that logic was in the forking pr, not in master, nevermind
```

**Comment by pgarg66** [comment]
@garious I captured the discord discussion in the comment above. I can use help in identifying the direction on this issue. Do we want to rework replay stage as part of this, or file a separate issue for that?

**Comment by garious** [comment]
Do what it takes. Do not get blocked. Personally, I'd guess replay stage needs rework to do a quality job here.  I think replay stage needs rework for a bunch of other issues too. I'll be taking a stab at reworking it in the next day or two, but don't wait on me.
